### PR TITLE
Make virtual network peerings reconcile/delete async

### LINKED ISSUE #1829: async vnet peerings

### PR DESCRIPTION
**Pull request by Jont828**
**What type of PR is this?**

/kind feature
**What this PR does / why we need it**: Implementation of an async service for virtual network peerings as a follow up for #1610 and #1541.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1829 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Make virtual network peerings reconcile/delete async
```

### CODE REVIEW

File: azure/services/vnetpeerings/vnetpeerings.go [hunk 1]
```diff
@@ -58,52 +61,45 @@ func (s *Service) Reconcile(ctx context.Context) error {
...
-			peeringSpec.SourceVnetName,
-			peeringSpec.PeeringName,
-			network.VirtualNetworkPeering{
-				VirtualNetworkPeeringPropertiesFormat: &peeringProperties,
-			},
-		)
-
-		if err != nil {
-			return errors.Wrapf(err, "failed to create peering %s in resource group %s", peeringSpec.PeeringName, s.Scope.Vnet().ResourceGroup)
+	// We go through the list of route tables to reconcile each one, independently of the result of the previous one.
```
Comment (CecileRobertMichon) [inline azure/services/vnetpeerings/vnetpeerings.go:67]: s/route tables/vnet peerings

File: api/v1beta1/conditions_consts.go [hunk 2]
```diff
@@ -82,6 +82,8 @@ const (
 	ResourceGroupReadyCondition clusterv1.ConditionType = "ResourceGroupReady"
 	// VNetReadyCondition means the virtual network exists and is ready to be used.
 	VNetReadyCondition clusterv1.ConditionType = "VNetReady"
+	// VnetPeeringReadyCondition means the virtual network peerings exist and are ready to be used.
+	VnetPeeringReadyCondition clusterv1.ConditionType = "VnetPeeringReady"
```
Comment (CecileRobertMichon) [inline api/v1beta1/conditions_consts.go:86]: let's add this new condition to the list of summary and owned conditions in cluster.go (search the code for ResourceGroupReadyCondition to see where)

File: azure/services/vnetpeerings/client.go [hunk 3]
```diff
@@ -61,36 +68,135 @@ func (ac *AzureClient) Get(ctx context.Context, resourceGroupName, vnetName, pee
...
 	defer span.End()
 
-	future, err := ac.peerings.CreateOrUpdate(ctx, resourceGroupName, vnetName, peeringName, peering, network.SyncRemoteAddressSpaceTrue)
+	peering, err := ac.VnetPeeringParams(ctx, spec)
 	if err != nil {
-		return err
+		return nil, nil, errors.Wrapf(err, "failed to get desired parameters for virtual network peering %s", spec.ResourceName())
+	} else if peering == nil {
+		// nothing to do here.
+		return nil, nil, nil
```
Comment (CecileRobertMichon) [inline azure/services/vnetpeerings/client.go:83]: I think we'll also want to do https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1837/commits/aee609ef6e1268d28b6f9a49c6b56a9cb9888875 after I figure out why the tests are failing in that pr

File: azure/services/vnetpeerings/client.go [hunk 4]
```diff
@@ -61,36 +68,135 @@ func (ac *AzureClient) Get(ctx context.Context, resourceGroupName, vnetName, pee
...
 }
 
-// Delete deletes the specified virtual network peering.
-func (ac *AzureClient) Delete(ctx context.Context, resourceGroupName, vnetName, peeringName string) error {
-	ctx, span := tele.Tracer().Start(ctx, "vnetpeerings.AzureClient.Delete")
+// DeleteAsync deletes a route table asynchronously. DeleteAsync sends a DELETE
+// request to Azure and if accepted without error, the func will return a Future which can be used to track the ongoing
+// progress of the operation.
+func (ac *AzureClient) DeleteAsync(ctx context.Context, spec azure.ResourceSpecGetter) (azureautorest.FutureAPI, error) {
+	ctx, span := tele.Tracer().Start(ctx, "routetables.AzureClient.Delete")
```
Comment (Jont828) [inline azure/services/vnetpeerings/client.go:110]: TODO: fix name `routetables` -> `vnetpeerings`

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/assign @mboersma

**Review by mboersma — APPROVED** [review]
/lgtm

**Comment by CecileRobertMichon** [comment]
One thing that @Jont828 and I discussed offline that we should look into is how we can omit conditions when they don't apply to a specific AzureCluster. For example, if we set `VnetPeeringsReady` to `true`, it implies that there are vnet peerings in the cluster and that they are ready. However, currently, that is not the case since if there are 0 resources of that type, the condition will still be set to true so that it doesn't block the overall `Ready` status. @Jont828 it would be good to open an issue for this so we can brainstorm ideas.

/approve

**Pull request merged**